### PR TITLE
Locker based combat, evolved.

### PR DESCRIPTION
**Pull request by ArcaneMusic**
## About The Pull Request

You're walking down the halls of /tg/station 13. It is a good day. You had yourself a nice plate of eggs, bacon, and a crispy biscuit before you logged on.  You sign on as a chief medical officer. Today will be a good day to play on [TERRY/BAGIL/SYBIL/MANUEL]. You get your trusty hypospray and look through your locker. You close the locker, but forget to lock it.

Suddenly, tragedy strikes. You hear a sound, a familiar one. A screwdriver hitting airlock. You turn to your left, no, it's not from Medbay. It's coming from maintenance. You hear a Villain, a vile cretin, on the other side of the door. You think to message security, but they're all in the brig still. You panic. You ask a doctor to get help. No response. After all, this is [TERRY/BAGIL/SYBIL/MANUEL], after all. You feel sick to your real stomach. All those eggs and bacon are no longer sitting well knowing that you are about to be mugged by, no doubt, some kind of vile troublemaker.

Thankfully, an idea strikes. A golden idea. Suddenly, your fears leave your body. You move your locker near to the door. The thief will surely be distracted by all these free goodies. A final multitool snap pierces the air. A gray clad man walks into the office, his gaze blocked by a cheap, rubber gas mask produced by Nanotrasen, all too familiar to the minimum wage crew of this station. He sees you, grinning, but curiously not brandishing his telescopic baton. The assailant quickly makes a break for the locker, opening it up wildly in a mad dash to take some kind of expensive goodie as a trophy from this assault.

"Heh."

You do a forward thrusted shoulder charge, disarming the target and knocking them single handedly into your locker. Now trapped with his potential prize, he begins screaming at you, asking what trickery is this. You make a single deft swipe of your ID card in the air, and the microelectronics block the assistant in, trapped until further help could be found.

You appreciate having a good breakfast.

https://user-images.githubusercontent.com/41715314/124367536-007d1f00-dc26-11eb-9d6f-f7349babd9cd.mp4

## Why It's Good For The Game

It adds an interesting level of depth to shove combat, as before this you could already shove as a combo with lockers, but far less dynamically. Before, you could shove someone into a locker and then open it, then shove them inside again, and then close it. This is a more dynamic upgrade to that action, which would exchange chain-stunning someone with more or less capturing them inside, giving them a chance to potentially get back up.

Plus, you gain the ability to shove people into lockers as a combat action.

## Changelog
:cl:
add: You can now disarm someone into an available open locker to close it on them as your disarming action.
add: You can now disarm someone into an available closed locker to open the locker behind them up.
/:cl:

### CODE REVIEW

**Comment by JJRcop** [comment]
👏👏👏👏👏👏👏👏👏👏

**Comment by RandomGamer123** [comment]
an actual mechanic for 60000 for once

**Comment by Bawhoppen** [comment]
Ah! Nice get!

For the actual feature, maybe it should only close when you shove them into it if it's against the wall? Like cause the force hitting the back of the locker being blocked causes the door to shut closed in reaction.

File: code/modules/mob/living/carbon/carbon_defense.dm [hunk 1]
```diff
@@ -260,10 +260,12 @@
 	var/mob/living/carbon/target_collateral_carbon
 	var/obj/structure/table/target_table
 	var/obj/machinery/disposal/bin/target_disposal_bin
+	var/obj/structure/closet/target_storage_container
```
Comment (Mothblocks) [inline code/modules/mob/living/carbon/carbon_defense.dm:263]: This is sort of getting ridiculous, make this signal based (and for brownie points, make everything else signals too).

File: code/modules/mob/living/carbon/carbon_defense.dm [hunk 2]
```diff
@@ -260,10 +260,12 @@
...
 	var/shove_blocked = FALSE //Used to check if a shove is blocked so that if it is knockdown logic can be applied
 
 	//Thank you based whoneedsspace
 	target_collateral_carbon = locate(/mob/living/carbon) in target_shove_turf.contents
+	target_storage_container = locate(/obj/structure/closet) in target_shove_turf.contents
```
Comment (Mothblocks) [inline code/modules/mob/living/carbon/carbon_defense.dm:268]: This is part of why this needs to be signal based.

We're performing *two* O(n) searches for *every* shove. The more we add, the slower this is going to get.

File: code/modules/mob/living/carbon/carbon_defense.dm [hunk 3]
```diff
@@ -327,6 +329,26 @@
...
 			log_combat(src, target, "shoved", "into [target_disposal_bin] (disposal bin)")
+		else if(target_storage_container)
+			target.Knockdown(SHOVE_KNOCKDOWN_SOLID)
+			if(target_storage_container.opened)
+				target.forceMove(target_storage_container)
+			if(!(target_storage_container.locked || target_storage_container.welded))
+				target_storage_container.toggle()
+			target_storage_container.update_icon()
+			target.visible_message(span_danger("[name] shoves [target.name] into \the [target_storage_container]!"),
+							span_userdanger("You're shoved into \the [target_storage_container] by [target.name]!"), span_hear("You hear aggressive shuffling followed by a loud thud!"), COMBAT_MESSAGE_RANGE, src)
```
Comment (Mothblocks) [inline code/modules/mob/living/carbon/carbon_defense.dm:340]: Only go up one indentation level.

File: code/modules/mob/living/carbon/carbon_defense.dm [hunk 4]
```diff
@@ -327,6 +329,26 @@
...
+			to_chat(src, span_danger("You shove [target.name] into \the [target_storage_container]!"))
+			log_combat(src, target, "shoved", "into [target_storage_container] (locker/crate)")
+	else if (target_storage_container && !target.buckled)
+		//We are shoving a target into an available, unobstructed locker.
+		if(!(target_storage_container.locked || target_storage_container.welded))
+			target_storage_container.toggle()
+			target_storage_container.update_icon()
+		target.visible_message(span_danger("[name] shoves [target.name] into \the [target_storage_container]!"),
+				span_userdanger("You're shoved into \the [target_storage_container] by [target.name]!"), span_hear("You hear aggressive shuffling followed by a loud thud!"), COMBAT_MESSAGE_RANGE, src)
```
Comment (Mothblocks) [inline code/modules/mob/living/carbon/carbon_defense.dm:349]: Same here.

**Review by Mothblocks — CHANGES_REQUESTED** [review]

**Comment by ArcaneMusic** [comment]
Running into a funny snag right now with base signal issues, I still don't have a foggy idea on how to exactly set it up with connect loc but I can't do that if I can't even send the base signal I suppose.

**Comment by JJRcop** [comment]
Is this the real meme get?

**Comment by Cyberboss** [comment]
@ArcaneMusic finish this

**Comment by tralezab** [comment]
image
lol

File: code/game/objects/structures/crates_lockers/closets.dm [hunk 5]
```diff
@@ -676,4 +679,18 @@
 /obj/structure/closet/return_temperature()
 	return
 
+/obj/structure/closet/proc/locker_carbon(obj/structure/closet/closet, mob/living/carbon/shover, mob/living/carbon/target)
+	SIGNAL_HANDLER
+	//target.Knockdown(SHOVE_KNOCKDOWN_SOLID)
```
Comment (tralezab) [inline code/game/objects/structures/crates_lockers/closets.dm:684]: remove this commented code!

**Review by tralezab — CHANGES_REQUESTED** [review]
boop

**Comment by tralezab** [comment]
scared?

**Comment by JJRcop** [comment]
The get that didn't then got again (twice)

**Comment by Cyberboss** [comment]
We don't get got again

**Comment by kopoba** [comment]
cursed

**Pull request merged**